From 3b7ba39fcaf261f800205fd6cf45fea9998529e5 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Mon, 14 Jan 2013 14:03:54 +1000
Subject: [PATCH evdev 1/4] Localise tmp variable

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 src/evdev.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/evdev.c b/src/evdev.c
index 5667dc1..dfce8c4 100644
--- a/src/evdev.c
+++ b/src/evdev.c
@@ -424,7 +424,6 @@ EvdevQueueButtonClicks(InputInfoPtr pInfo, int button, int count)
 static void
 EvdevProcessValuators(InputInfoPtr pInfo)
 {
-    int tmp;
     EvdevPtr pEvdev = pInfo->private;
     int *delta = pEvdev->delta;

@@ -456,6 +455,7 @@ EvdevProcessValuators(InputInfoPtr pInfo)
     }

     if (pEvdev->rel_queued) {
+        int tmp;
         int i;

         if (pEvdev->swap_axes) {
--
1.8.1
